sui_programmability/framework/sources/Governance/Validator.move

```diff
@@ -0,0 +1,212 @@
+        ctx: &mut TxContext,
+    ) {
+        let sender = TxContext::sender(ctx);
+        assert!(self.sui_address == sender, EADDRESS_MISMATCH);
+
+        let Validator { id, sui_address: _, name: _, ip_address: _, stake, pending_stake, pending_withdraw } = self;
+        assert!(pending_withdraw == 0, EINCONSISTENT_STATE);
+        Transfer::transfer(stake, sender);
+        ID::delete(id);
+        Option::destroy_none(pending_stake);
+    }
+
+    public(script) fun add_stake(
+        self: &mut Validator,
+        new_stake: Coin<SUI>,
+        _ctx: &mut TxContext,
+    ) {
+        Coin::join(&mut self.stake, new_stake)
+    }
+
+    public(script) fun withdraw_stake(
+        self: &mut Validator,
+        withdraw_amount: u64,
+        ctx: &mut TxContext,
+    ) {
+        let coin = Coin::withdraw(&mut self.stake, withdraw_amount, ctx);
+        Coin::transfer(coin, TxContext::sender(ctx))
+    }
+
+
+    public(friend) fun new(
+        sui_address: address,
+        name: vector<u8>,
+        ip_address: u128,
+        stake: Coin<SUI>,
+        ctx: &mut TxContext,
+    ): Validator {
+        Validator {
+            id: TxContext::new_id(ctx),
+            sui_address,
+            name: ASCII::string(name),
+            ip_address,
+            stake,
+            pending_stake: Option::none(),
+            pending_withdraw: 0,
+        }
+    }
+
+    /// Called by `ValidatorSet`, to send back a Validator object to its address.
+    /// This happens when a validator is withdrawn or does not quality.
+    public(friend) fun send_back(self: Validator) {
+        check_non_active_validator_invariants(&self);
+        let owner = self.sui_address;
+        Transfer::transfer(self, owner)
+    }
+
+    /// Add stake to a pending validator. Since the validator is pending,
+    /// we can add the stake directly to its stake.
+    public(friend) fun add_stake_to_pending_validator(
+        self: &mut Validator,
+        new_stake: Coin<SUI>,
+    ) {
+        check_non_active_validator_invariants(self);
+        Coin::join(&mut self.stake, new_stake)
+    }
+
+    /// Add stake to an active validator. Since it's active, we will need to
+    /// add the stake to the pending_stake field, which will be processed at the
+    /// end of epoch.
+    public(friend) fun request_add_stake_to_active_validator(
+        self: &mut Validator,
+        new_stake: Coin<SUI>,
+    ) {
+        if (Option::is_none(&self.pending_stake)) {
+            Option::fill(&mut self.pending_stake, new_stake)
+        } else {
+            let pending_stake = Option::extract(&mut self.pending_stake);
+            Coin::join(&mut pending_stake, new_stake);
+            Option::fill(&mut self.pending_stake, pending_stake);
+        }
+    }
+
+    /// Withdraw stake from a pending validator. Since it's pending, we can
+    /// withdraw coins immediately and send it back to the validator's address.
+    public(friend) fun withdraw_stake_from_pending_validator(
```

Comment:
I think we should either not allow this (slightly preferred, since it's a bit simpler), or should only allow it if the amount of stake remaining is still above the minimum. Otherwise, someone can spam the pending validator set by adding themselves, then withdrawing all of their stake. We must try to ensure that the pending set doesn't get too big (e.g., hit the object size limit).